### ISSUE TITLE
Docs: --kv-mean-center workflow for q4_0 K-cache (needed for 262k ctx on 16 GB cards)

### ISSUE THREAD

**Issue by Astezelex** [PrismML-Eng/llama.cpp#85]
Context: we ran what we believe is the first independent benchmark of Ternary Bonsai 27B on consumer hardware (RTX 5060 Ti 16 GB, fork commit b9591, CUDA build for sm_120, Linux). Overall results were good for the fork: ternary +24-31% decode vs a best-engine IQ2_XXS baseline, 262k context fits a 16 GB card with q4_0 KV, and the DSpark lossless-precision claim verified byte-identical Q4_1 vs bf16 (5/5 prompts at temp 0). Raw data and scripts: https://github.com/Astezelex/bonsai-27b-16gb-bench (also PrismML-Eng/Bonsai-demo#98). This report is friction we hit on the way.

The fork docs say a q4_0 K-cache "wants" a `--kv-mean-center` bias file for quality, but there is no guidance on generating or obtaining one per model, and no shipped file for Bonsai 27B. At 262k context (the fit headline for 16 GB cards) q4 KV is mandatory, so the quality delta matters.

Could you document the workflow (how to produce the bias file for a given model), or state that it is safe to omit for Bonsai 27B?

**Comment by khosravipasha** [comment]
Nice, this is impressive.

We did not have time to provide one, can you check if this guide is sufficient for now to unblock you? https://github.com/PrismML-Eng/llama.cpp/issues/54

Will take a look next week on our side to upload a sample centering gguf.

You can try other kv-cache quantization methods as well, this one is an experimental one that gave us early good results but its not fully integrated in hte llama.cpp fork yet, we will polish it more as we go forward.

**Comment by Astezelex** [comment]
Thanks, yes, #54 unblocks us. Our tested commit (b9591) already carries `tools/kv-mean-center` in the tree; we had only built the bench/cli/server targets, so it's just an extra build target plus the calibration run. One clarification for anyone cross-reading with PR PrismML-Eng/Bonsai-demo#98: our published 262k-fit number was measured with plain q4_0 KV, no centering bias, so it stands independent of this feature. A vendor sample centering GGUF for Bonsai 27B would be the ideal ending for the 16 GB story: at 262k the q4 K-cache is mandatory, so the bias file is effectively part of the deployment recipe. Happy to test the workflow on sm_120 and report back here.

**Comment by Astezelex** [comment]
Small addition useful for anyone landing here: the latest prebuilt release (prism-b9594-38c66ad) already ships `llama-kv-mean-center` in the linux-cuda tarball, so the #54 calibration workflow needs no source build. We'll report our Bonsai-27B calibration results on sm_120 here.

**Comment by khosravipasha** [comment]
Following up on this one, and genuinely curious about the results you mentioned reporting back: how did the q4_0-K + mean-center calibration work out on the 5060 Ti in terms of quality (vs uncentered q4 KV at your 262k target) and speed? Yours is the most rigorous independent setup that has exercised this feature, so the numbers would be valuable to have on record either way.

A summary of what changed since your tested commit, because it is a lot:

- The `prism` branch is now rebased onto **current mainline llama.cpp** (b9591 was an early-July base). Releases from [prism-b10658](https://github.com/PrismML-Eng/llama.cpp/releases) onward are the new line.
- Model files changed with it: v7 builds read the official group-64 `*_Q2_0_g64.gguf` / 27B `*_Q2_g64.gguf` or the fork's `*-PQ2_0.gguf` (about 6% smaller, same weights, both exact re-encodings of the ternary weights); the old `*-Q2_0.gguf` is refused with a pointer.
- kv-mean-center survived the migration intact and is validated end to end, including together with speculative decoding. One new requirement: calibrate the bias with `-ctk q4_0` set so the K-rotation state matches the serving config (a mismatched bias is rejected at load with a clear message).
- **DSpark got faster and much smaller.** It now rides on mainline's draft-dspark plus fork patches; the published drafter needs a one-time conversion (`gguf-dspark-to-dflash --drop-shared-tensors` then `llama-quantize Q4_0`, exact commands in [SPECULATIVE.md](https://github.com/PrismML-Eng/Bonsai-demo/blob/main/SPECULATIVE.md)), after which it is ~0.6 GB (vs 1.9) and we measure 27B decode at **+73 to +93% over no-draft on CUDA** on our hardware. Your byte-identical-at-temp-0 property still holds by construction. On a 16 GB card the 1.3 GB of VRAM the slim drafter returns is basically free context.
- Since you build for sm_120: native CUDA 13.3 builds (Linux and Windows) are landing in the release matrix, which should remove your need to build from source at all.

If you are willing to re-run your benchmark suite against the new release, we would love to see the deltas, especially the ternary-vs-IQ2_XXS comparison now that the fork's non-ternary paths are identical to mainline (that 4.5% skew from #83 is gone by construction), and the dspark speedup with the slim drafter at your context sizes. Keeping this open for your results.

**Comment by Astezelex** [comment]
Thanks for the detailed summary, and sorry for the slow reply. I re-ran the suite against
b10658 with `Q2_g64`, plus the two things you asked about directly.

Full write-up, with every command, the figures and the raw artefacts each number comes from:
**https://github.com/Astezelex/bonsai-27b-16gb-depth**
Part 1, for context: https://github.com/Astezelex/bonsai-27b-16gb-bench

The short version is below.

**Most of what you listed reproduces, including two things that were broken in July.**
`Q2_g64` no longer produces gibberish and `PQ2_0` no longer fails with `invalid ggml type 142`;
both now load and generate coherently, checked for non-empty output and not just a zero exit.
The rebase itself cost nothing: decode went 44.4 to 44.74 tok/s, +0.8%, inside run-to-run
noise. The rebased release line works on sm_120 with no
source build, `Q2_g64` loads and the old `Q2_0` is refused with a clear pointer, `PQ2_0` is
5.54% smaller (7,165,121,600 B against 7,585,330,240 B) and also slightly *faster* (46.02
against 44.35 tok/s decode), the bias does have to be calibrated with `-ctk q4_0`, centering
works together with speculative decoding, and the slim drafter is 602 MiB and gives
**+85.0%** at depth 0, inside your +73% to +93% band. At 8k context it gives **+104.2%**.

Five things did not come out the way the summary describes. Each has a suggested fix.

---

**1. Mean-centering quality, which was your original question.**

The published `-24.6%` mean KLD reproduced exactly (uncentered 0.000646, centered 0.000487).
Your docs report `-22.9%` for this configuration on a different model, so on different hardware
and against a held-out split this is an independent replication of the effect size and not just
of the sign. Three caveats belong beside it: the **maximum** KLD is marginally worse centered
(0.0253 against 0.0237), so the tail is not improved and only the bulk is; the sample is 12
chunks; and KLD says the cached keys are closer to the F16 reference, not that a benchmark
score moves.

Two things change it.

The published measurement sets `-ctk q4_0` and leaves V at f16. That asymmetric pair falls off
the fused FA kernel: on my hardware `-ctk q4_0 -ctv f16` runs prefill at 241 tok/s against
1,020 for the symmetric `q4_0/q4_0`, and at depth 8,192 it is **4.53 against 907.71**. Nobody
would serve that way. In the symmetric configuration a 262k deployment needs, the benefit at
512 tokens is **-7.3%**.

Then it decays with context: **-9.5%** at 512, **-0.7%** at 8,192, **+0.7%** at 16,384.

I could not take this to 262k, and the reason may interest you more than the number:
`llama-perplexity` reserves `n_ctx x n_vocab x 4` bytes of host RAM, which is 159.3 GB at
262,144. My 32 GB host is OOM-killed at `n_ctx 32768`. **The quality of a feature built for
262k context cannot currently be validated at 262k by the tool that ships beside it.** A
streaming or chunked logits path would fix that.

On speed, the other half of your question: centering is free. Across six depths from 0 to
262,144 it moves decode by between -0.24% and +0.30%, which straddles zero, and it costs
**exactly 2 MiB of VRAM at every one of them, including 262,144**.

**2. The slim drafter is a 4,096-token model.**

Acceptance against context depth, held-out Wikipedia prefix, wall-clock cross-checked:

| depth | acceptance | uplift |
|---|---|---|
| 0 | 59.2% | +85.0% |
| 8,208 | 49.5% | +104.2% |
| 16,488 | 18.1% | +40.7% |
| 32,918 | 0.085% | +0.4% |

Your own log says why:

```
llama_context: n_ctx_seq (135168) > n_ctx_train (4096) -- possible training context overflow
```

`dflash-slim-Q4_0.gguf` carries `n_ctx_train = 4096`. Acceptance holds near that window and
is gone at eight times it.

Against the drafter I used in Part 1 the conversion is a clear win in both directions:
1,946,393,568 B to **631,712,480 B**, and 79.1 to 82.31 tok/s, so +78.2% to **+85.8%** uplift.
Smaller and faster at once. The cold first repetition read 79.99 against 82.30 and 82.31 warm;
the warm figure is the one quoted.

Worth stressing, because it cuts in your favour: **prefix content matters as much as length.**
On a real workload where the long context is the model's own generated reasoning, the same
drafter holds **18% to 31% acceptance at 60k to 130k generated tokens**. It collapses only on
an unfamiliar prefix. And it costs nothing in quality: AIME26 with and without the drafter
gives 26/30 against 27/30, identical cap counts, accuracy-when-converged 0.9630 against
0.9643.

**3. The 1.3 GB the slim drafter returns is not free context.**

The drafter allocates its own KV cache, so it costs far more than its file size:

| context | no draft | drafted | drafter costs |
|---|---|---|---|
| 8,192 | 7,893 MiB | 10,629 MiB | 2,736 MiB |
| 135,168 | 10,735 | 14,969 | **4,234** |
| 266,240 | 13,679 | **fails to load** | |

At 262k the drafted server does not start on a 16 GB card (`common_fit_params: failed to fit
params to free device memory`). I left the draft cache at its f16 default; `-ctkd`/`-ctvd`
should recover much of this and I have not measured it.

**4. A mismatched bias is not rejected.**

A truncated bias file is rejected cleanly. But a bias **calibrated at `-ctk q8_0` and served
at `-ctk q4_0`** loads without complaint, and the files genuinely differ (different SHA-256,
35,409 of 66,400 bytes). The reason is in the format: a bias GGUF carries only

```
general.type          str  = kv-mean-center
kv_mean_center.k_rot  bool = 1
```

There is no key recording the cache type it was calibrated with, so the loader has nothing to
compare. A `kv_mean_center.cache_type_k` key plus a check at load would close it.

**5. The byte-identical-at-temp-0 property does not hold.**

This is the one I would most want you to look at. Five fixed prompts, temp 0, top_k 1, seed
1234, `max_tokens 900` so they finish naturally, compared by SHA-256 of the exact string with
`reasoning_content` and `content` concatenated:

| `--spec-draft-n-max` | identical to greedy |
|---|---|
| 1 | **3/5** |
| 2 | **3/5** |
| 4 | **2/5** |

Controls first: greedy against greedy is 5/5 identical in the same process and 5/5 across a
fresh process, and each drafted configuration is 5/5 self-consistent. So the engine is
deterministic and the difference is systematic.

**Divergence happens even at block size 1.** If it only appeared at larger blocks I would
suspect accumulation across the block. At n-max=1 the suspect is the verify-and-accept step.
And the prompts that diverge have consistently lower draft acceptance than the ones that stay
bit-exact, with the gap widening as the block grows (+0.08, +0.10, +0.18), so divergence tracks
rejections and not acceptances. That points at the rollback path.
The divergences are fluent alternatives and not corruption, and one appears at character 1,487,
so it is not confined to the opening tokens. What this is not: not a rate (3/5 has a Wilson
interval of [23%, 88%]), not a significance test, and not a quality claim.

Related, and a correction to my own Part 1: I briefly suspected my July 5/5 identity result for
the older `Q4_1` drafter was a harness artefact. It was not, and re-measuring the archived
generations confirms 5/5. One limit I should have stated then: all fifteen of those generations
hit `finish_reason=length` at 512 tokens, so that was identity over a 512-token prefix. The
divergence above shows at character 1,487, so the old drafter was never tested at the length
where this gets stressed.

**6. The 4.5% skew is halved, not gone.**

Interleaved A/B/A/B on the identical IQ2_XXS file, one card: fork b10658 averages 34.42 tok/s
decode, mainline b10588 averages 35.28. That is **2.42%**, down from July's 4.47%, and the
narrowing came mostly from mainline moving down. Caveat: seventy builds apart, so close to
like-for-like and not exactly it.

Also, the release matrix I can see carries CUDA 12.4 and 12.8, no 13.3 yet.

---

**One thing I cannot explain**, in case it means something to you. At a 131k prefix the
drafted server produced 300 tokens in 31 s against 45 s undrafted, reproducible across two
runs with different prefix content, while the acceptance counter read `0 accepted / 1186
generated, mean len = 1.00`. Server logs show no configuration difference between the arms. A
50% speedup with zero accepted draft tokens is not something I can account for.

On the ternary-against-IQ2_XXS comparison you asked for, the rebase removed a confound Part 1
could not. In July each model had to be given its best engine. Both now run on one binary:

| same binary b10658, one card | size | pp512 | tg128 |
|---|---|---|---|
| Bonsai 27B `Q2_g64` | 7.05 GiB | 1046.85 | **44.74** |
| Qwen3.6-27B `UD-IQ2_XXS` | 8.73 GiB | 997.20 | 35.05 |

Ternary is **27.6% faster at decode and 1.68 GiB smaller** with no engine difference left to
explain it away. Part 1's cross-engine figures were 44.4 against 35.8, so the advantage was if
anything understated before.

On quality: the Part 1 finding holds on b10658.
The gap is convergence, not knowledge. On the 18 AIME items both models completed, both scored
18/18 in both runs; what separates them is that IQ2_XXS fails to finish 11 of 30 within a 60k
budget against Bonsai's 3, and those cap counts are identical to July's.

Happy to re-run any of this if you want a different configuration, and happy to open separate
issues for the bias metadata key and the perplexity memory ceiling if that is more useful than
one thread.

Chapter references above point into https://github.com/Astezelex/bonsai-27b-16gb-depth.

**Comment by zhaoyilun** [comment]
Adding a data point from the "no bias at all" side, since it may narrow the scope of the doc request.

On RTX 4080 SUPER 16 GB (Ada, CUDA 13.3), `-c 262144 -ctk q4_0 -ctv q4_0` runs **without any `--kv-mean-center` bias**, and what actually broke 256k on this card was the K/V type, not the context size:

| config (`-c 262144`) | prefill (12.5k prompt) | decode | VRAM |
|---|---:|---:|---:|
| PQ2_0 + MTP, `-ctk q8_0 -ctv q8_0` | **101 -> 35 t/s** (one 10,240-token prefill: 293 s) | — | 15.7 GiB |
| PQ2_0 + MTP, `-ctk q4_0 -ctv q4_0` | **1726 t/s** | 90.4 t/s median (MTP) | 15.7 GiB |

Two notes that might save someone time:

1. `common_fit_params: failed to fit params to free device memory` shows up in the **q4_0** state too, while performance is completely fine — so that warning is not a usable signal. KV cache bytes are.
2. Long-context measurement is easy to get wrong: `timings.prompt_n` reports only the **newly evaluated** tokens, because context checkpoints are reused across requests (a 191k-token prompt came back as `prompt_n = 63745`). The server is not truncating — its log shows `n_tokens = 191139, truncated = 0`. Interleaving a short "cache buster" request does **not** evict the checkpoints; use `POST /tokenize` for the real length, or sum the incremental `prompt_ms` across a cumulative ladder.

Depth curve and raw JSON: <https://github.com/zhaoyilun/bonsai2-27b-mtp-repro> (decode falls 86.9 -> 35.0 t/s from 8k to 191k with MTP; 63.0 -> 26.5 without). Also worth linking @Astezelex's independent 16 GB write-up from this thread — two data points on the same "does 262k really fit a 16 GB card" question.
